### PR TITLE
Update delivery on detail change

### PR DESCRIPTION
**Pull request by RossStewart**
C# changes. Question is now do we remove recalculateDetailFields from the purchaseorder reducer? Going to submit that as a separate PR if consensus is yes

### CODE REVIEW

**Comment by richardiphillips** [comment]
> C# changes. Question is now do we remove recalculateDetailFields from the purchaseorder reducer? Going to submit that as a separate PR if consensus is yes

why would we remove it?

File: src/Domain.LinnApps/PurchaseOrders/PurchaseOrderService.cs [hunk 1]
```diff
@@ -958,8 +958,20 @@ private void SetDetailFieldsForCreation(PurchaseOrderDetail detail, int newOrder
         {
             foreach (var delivery in deliveries)
             {
-                var updatedDelivery = updatedDeliveries.First(x => x.DeliverySeq == delivery.DeliverySeq);
-                delivery.DateRequested = updatedDelivery.DateRequested;
+                if (delivery.QuantityOutstanding > 0)
+                {
+                    var updatedDelivery = updatedDeliveries.First(x => x.DeliverySeq == delivery.DeliverySeq);
```
Comment (lewisrenfrew) [inline src/Domain.LinnApps/PurchaseOrders/PurchaseOrderService.cs:963]: If I'm reading this right it looks like you are just setting all the prices based off of what the delivery price used to be. How is that going to keep them in sync with any changes to the prices on the order detail?

File: src/Domain.LinnApps/PurchaseOrders/PurchaseOrderService.cs [hunk 2]
```diff
@@ -958,8 +958,20 @@ private void SetDetailFieldsForCreation(PurchaseOrderDetail detail, int newOrder
...
+                    delivery.OurUnitPriceCurrency = updatedDelivery.OurUnitPriceCurrency;
+                    delivery.NetTotalCurrency = Math.Round((decimal)(updatedDelivery.OurDeliveryQty * updatedDelivery.OurUnitPriceCurrency));
+                    delivery.VatTotalCurrency = updatedDelivery.VatTotalCurrency;
+                    delivery.DeliveryTotalCurrency = updatedDelivery.DeliveryTotalCurrency;
+                    delivery.BaseOurUnitPrice = updatedDelivery.BaseOurUnitPrice;
+                    delivery.BaseOrderUnitPrice = updatedDelivery.BaseOrderUnitPrice;
+                    delivery.BaseVatTotal = updatedDelivery.BaseVatTotal;
+                    delivery.BaseDeliveryTotal = Math.Round((decimal)((updatedDelivery.OurDeliveryQty * updatedDelivery.BaseOurUnitPrice.GetValueOrDefault()) + updatedDelivery.BaseVatTotal), 2);
+                    delivery.BaseNetTotal = Math.Round((decimal)(updatedDelivery.OurDeliveryQty * updatedDelivery.BaseOurUnitPrice.GetValueOrDefault()), 2);
+                }
```
Comment (lewisrenfrew) [inline src/Domain.LinnApps/PurchaseOrders/PurchaseOrderService.cs:974]: How are you calculating VAT etc?

File: tests/Unit/Domain.LinnApps.Tests/PurchaseOrderServiceTests/WhenUpdatingPurchaseOrderConvValueQtyFields.cs [hunk 3]
```diff
@@ -59,24 +60,24 @@ public void SetUp()
...
-                                                                           OrderUnitPriceCurrency = 120m,
+                                                                           OurUnitPriceCurrency = 0m,
+                                                                           OrderUnitPriceCurrency = 0m,
                                                                            BaseOrderUnitPrice = 100m,
                                                                            VatTotalCurrency = 0m,
                                                                            BaseVatTotal = 0m,
-                                                                           DeliveryTotalCurrency = 120m,
-                                                                           BaseDeliveryTotal = 100m,
+                                                                           DeliveryTotalCurrency =  0m,
+                                                                           BaseDeliveryTotal = 0m,
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.LinnApps.Tests/PurchaseOrderServiceTests/WhenUpdatingPurchaseOrderConvValueQtyFields.cs:80]: why set all these to 0?

File: tests/Unit/Domain.LinnApps.Tests/PurchaseOrderServiceTests/WhenUpdatingPurchaseOrderConvValueQtyFields.cs [hunk 4]
```diff
@@ -395,5 +392,35 @@ public void ShouldUpdateMiniOrderFields()
...
+        {
+            this.current.OrderNumber.Should().Be(600179);
+            this.current.Remarks.Should().Be("updated remarks");
+
+            var firstDetail = this.current.Details.First();
+
+            var delivery = firstDetail.PurchaseDeliveries.First();
+
+            // order unit price currency, our unit price currency, our delivery qty
+            delivery.OrderUnitPriceCurrency.Should().Be(120m);
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.LinnApps.Tests/PurchaseOrderServiceTests/WhenUpdatingPurchaseOrderConvValueQtyFields.cs:407]: You didn't change this away from 120m in these tests so this isn't testing the delivery price was updated

File: tests/Unit/Domain.LinnApps.Tests/PurchaseOrderServiceTests/WhenUpdatingPurchaseOrderOverrideValueQtyFields.cs [hunk 5]
```diff
@@ -392,5 +388,35 @@ public void ShouldUpdateMiniOrderFields()
...
+        {
+            this.current.OrderNumber.Should().Be(600179);
+            this.current.Remarks.Should().Be("updated remarks");
+
+            var firstDetail = this.current.Details.First();
+
+            var delivery = firstDetail.PurchaseDeliveries.First();
+
+            // order unit price currency, our unit price currency, our delivery qty
+            delivery.OrderUnitPriceCurrency.Should().Be(120m);
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.LinnApps.Tests/PurchaseOrderServiceTests/WhenUpdatingPurchaseOrderOverrideValueQtyFields.cs:403]: this test updated this to be 200.22 on the detail - Why are you testing that its 120 here?

**Review by lewisrenfrew — CHANGES_REQUESTED** [review]
Looks like you've misunderstood the issue here - you need to update prices on all outstanding deliveries to be the new prices entered by the user

**Review by lewisrenfrew — APPROVED** [review]
looks good - Should work out / ask Andy & co what should happen when they change the qty next. presumably the deliveries will need updated in that case too.

(merge this first of course)

**Comment by RossStewart** [comment]
> looks good - Should work out / ask Andy & co what should happen when they change the qty next. presumably the deliveries will need updated in that case too.
> 
> (merge this first of course)

Looks like Andy is on holiday until the 31st, should I maybe ask Scott or Mandy? Will look into that issue we spoke about on stand up for the moment (Andrew is giving me details) and then see what the next move is

**Pull request merged**